src/recon_buildblock/ML_estimate_component_based_normalisation.cxx

```diff
@@ -84,225 +127,274 @@ ML_estimate_component_based_normalisation(const std::string& out_filename_prefix
         }
     }
 
-  FanProjData model_fan_data;
-  FanProjData fan_data;
-  DetectorEfficiencies data_fan_sums(IndexRange2D(num_physical_rings, num_physical_detectors_per_ring));
-  DetectorEfficiencies efficiencies(IndexRange2D(num_physical_rings, num_physical_detectors_per_ring));
-
-  GeoData3D measured_geo_data(num_physical_axial_crystals_per_basic_unit,
-                              num_physical_transaxial_crystals_per_basic_unit / 2,
-                              num_physical_rings,
-                              num_physical_detectors_per_ring); // inputes have to be modified
-  GeoData3D norm_geo_data(num_physical_axial_crystals_per_basic_unit,
-                          num_physical_transaxial_crystals_per_basic_unit / 2,
-                          num_physical_rings,
-                          num_physical_detectors_per_ring); // inputes have to be modified
-
-  BlockData3D measured_block_data(num_axial_blocks, num_transaxial_blocks, num_axial_blocks - 1, num_transaxial_blocks - 1);
-  BlockData3D norm_block_data(num_axial_blocks, num_transaxial_blocks, num_axial_blocks - 1, num_transaxial_blocks - 1);
-
-  make_fan_data_remove_gaps(model_fan_data, model_data);
-  {
-    // next could be local if KL is not computed below
-    FanProjData measured_fan_data;
-    float threshold_for_KL;
-    // compute factors dependent on the data
+  // Setup the data structures given the PET scanner geometry
+  data_fan_sums = DetectorEfficiencies(IndexRange2D(num_physical_rings, num_physical_detectors_per_ring));
+  efficiencies_ptr = std::make_shared<DetectorEfficiencies>(IndexRange2D(num_physical_rings, num_physical_detectors_per_ring));
+
+  measured_geo_data = GeoData3D(num_physical_axial_crystals_per_basic_unit,
+                                num_physical_transaxial_crystals_per_basic_unit / 2,
+                                num_physical_rings,
+                                num_physical_detectors_per_ring); // inputes have to be modified
+  norm_geo_data_ptr = std::make_shared<GeoData3D>(num_physical_axial_crystals_per_basic_unit,
+                                                  num_physical_transaxial_crystals_per_basic_unit / 2,
+                                                  num_physical_rings,
+                                                  num_physical_detectors_per_ring); // inputes have to be modified
+
+  measured_block_data = BlockData3D(num_axial_blocks, num_transaxial_blocks, num_axial_blocks - 1, num_transaxial_blocks - 1);
+  norm_block_data_ptr
+      = std::make_shared<BlockData3D>(num_axial_blocks, num_transaxial_blocks, num_axial_blocks - 1, num_transaxial_blocks - 1);
+
+  make_fan_data_remove_gaps(model_fan_data, *model_projdata_sptr);
+  make_fan_data_remove_gaps(measured_fan_data, *measured_projdata_sptr);
```

Comment:
My cpp-foo might not be up to snuff here. The target of the shared_ptr is being passed by reference into the function. Is this okay?

Reply:
as long as `make_fan_data_remove_gaps` doesn't store references, pointers or whatever to the content of `*measured_projdata_stpr`, then it is fine. (and it doesn't, as we'd have had compilation warnings previously anyway)

Reply:
`make_fan_data_remove_gaps` is purely functional.